Enable and disable power supply automatically

CReimer opened the pull request:
I use an ATX power supply for my printer and wanted it to only be powered on when there's actually a need for 12V. 

There's a new loop which checks for active steppers, active fans or active heaters.
Steppers are enabled instantly, to avoid missed steps.
Heaters and fans rely on the check() loop.

If nothing is left which needs 12V power PS_ON gets disabled after a predefined timeout.

Since this is my first contribution and I never developed for a microcontroller before I'm quite sure I did a lot wrong and there's a lot to improve.

CReimer committed:
Enable and disable power supply automatically

thinkyhead commented on Marlin/Marlin.h:
```diff
@@ -119,35 +120,35 @@ void manage_inactivity(bool ignore_stepper_queue = false);
 
 #if HAS_X2_ENABLE
-  #define  enable_x() do{ X_ENABLE_WRITE( X_ENABLE_ON); X2_ENABLE_WRITE( X_ENABLE_ON); }while(0)
+  #define  enable_x() do{ powerManager.power_on(); X_ENABLE_WRITE( X_ENABLE_ON); X2_ENABLE_WRITE( X_ENABLE_ON); }while(0)
```
We don't want to add this overhead to every stepper enable. And, if power was needed at this point, it might not turn on quickly enough to handle the immediately-following steps. It may require some milliseconds to "warm up."

thinkyhead commented on Marlin/Marlin.h:
```diff
@@ -119,35 +120,35 @@ void manage_inactivity(bool ignore_stepper_queue = false);
   #define disable_x() do{ X_ENABLE_WRITE(!X_ENABLE_ON); axis_known_position[X_AXIS] = false; }while(0)
 #else
-  #define  enable_x() NOOP
+  #define  enable_x() powerManager.power_on()
```
`NOOP` is used as a placeholder for a stepper that doesn't exist.

thinkyhead commented on Marlin/power.cpp:
```diff
@@ -0,0 +1,88 @@
+#if ENABLED(FAN_NEEDS_POWER)
+    for (uint8_t i = 0; i < FAN_COUNT; i++) {
+      if (fanSpeeds[i] > 0) {
+        powerNeeded = true;
```
It would be better to have a `is_power_needed` function (inline, perhaps) which has `return true` here. (Thus skipping all future tests.)

thinkyhead commented on Marlin/power.cpp:
```diff
@@ -0,0 +1,88 @@
+#if ENABLED(AUTO_FAN_NEEDS_POWER)
+    for (uint8_t i = 0; i < HOTENDS; i++) {
+      if (autoFanSpeeds[i] > 0) {
+        powerNeeded = true;
```
…`return true`.

thinkyhead commented on Marlin/power.cpp:
```diff
@@ -0,0 +1,88 @@
+      powerNeeded = true;
+    }
+
+    if (powerNeeded) {
```
`if (is_power_needed()) { . . .`

thinkyhead commented:
This is throwing compile errors.

The general idea is sound, but it might add some overhead for a condition which is easy to diagnose and handle manually.

We can't call into power-check in every stepper enable macro. These macros are used frequently in the planner and in the stepper ISR.

This PR would have to be incorporated as an optional feature also.

Plus, this PR modifies the expected behavior of some commands (i.e., causing `G1` to turn on the power supply, which is not in its specification), so it might not be merged on that basis.

CReimer committed:
Remove power_on signal from all stepper enables

CReimer committed:
Use a private function which returns a instead of dragging along a bool variable
Make sure POWER_TIMEOUT is always set to prevent compile errors
Only define controllerFanSpeed if there is a controller fan to prevent linker errors
Add Marlin's default copyright header

CReimer commented:
Thanks for the quick response. I just pushed a few changes to hopefully fix everything you mentioned above. Since we now no longer have an instant power up when a stepper gets activated I'd like to increase the loops speed. 2.5secs seem too long now.

CReimer committed:
Make power control optional

CReimer commented:
OK, after some more testing I'm not happy with relying on the loop for active stepper motor detection. At least we need something to enable power when some settings in the OSD are selected. e.g homing and manual moving.

CReimer committed:
Enable psu when there are any moves planned for a stepper
Make sure lastPowerOn timestamp get updated by external commands (M80)

CReimer commented:
The last change adds more or less the behavior of the Teacup firmware. The psu gets enabled when there is any move planned for a stepper.

Does this add much overhead? I don't know how much time OUT_WRITE() needs. I could keep track of the current PSU status and just return in Power::power_on() and Power::power_off() if the desired state is already there.

mtowara commented:
I have a very similar feature in my own Marlin fork, it would be very nice to have such a thing officially supported.

I like your changes, maybe you could make the timeout user customizable (I based my feature on M85).
Also I would suggest to add calls to power_on in setTargetHotend() and setTargetBed() (temperature.cpp) to avoid the max 2.5s delay which might trip up the temperature protection. Those functions should be called quite infrequently, so not much of a penalty there.

The OUT_WRITE() in power_on should be only a few cycles, the call to millis() is propably much more expensive. I think you can eliminate that and only check millis() in the check routine.

CReimer commented:
Sounds reasonable. But I'd like to hear back from @thinkyhead first. Because I'm afraid that any more changes here might delay the merge further.

CReimer committed:
Enable PSU before setting hotend or bed temperature to prevent problems with temperature protection
(Suggested by mtowara)

CReimer committed:
Do not set lastPowerOn time twice in the same check run

thinkyhead commented:
While we're focused on the 1.1 release, this will be on hold. We'll review it for the 1.1.1 update.

CReimer commented:
Well 1.1.1 is out for quite some time. We're actually already at 1.1.4.
Should I rebase this onto another branch or is there anything else I can do to speed things up?

thinkyhead commented:
Moved over to #9499
